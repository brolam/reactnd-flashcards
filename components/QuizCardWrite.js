import React from 'react'
import { StyleSheet, Text, Alert } from 'react-native'
import { View, TextInput, TouchableOpacity, KeyboardAvoidingView } from 'react-native'
import PropTypes from 'prop-types'
import cardStyle from '../styles/cardStyles'
import buttonStyle from '../styles/buttonStyles'
import { white, blue, green, red } from '../styles/colors'

export default function QuizCardWrite({
  quiz = { question: '', answer: '' },
  onSave = (question, answer, answerExpect) => { } }) {

  const defaultQuestion = quiz.question
  const defaultAnswer = quiz.answer
  let textInputQuestion = {}
  let textInputAnswer = {}

  function parseFields(answerExpect) {
    const question = textInputQuestion.value ? textInputQuestion.value : defaultQuestion
    const answer = textInputAnswer.value ? textInputAnswer.value : defaultAnswer
    if (!question || question.length == 0) {
      textInputQuestion.focus()
      Alert.alert('Question is required!')
      return
    }
    if (!answer || answer.length == 0) {
      textInputAnswer.focus()
      Alert.alert('Answer is required!')
      return
    }
    onSave(question, answer, answerExpect)
  }
  return (
    <KeyboardAvoidingView
      style={[cardStyle, styles.container]}
      behavior='padding'
    >
      <View style={styles.containerTextInput}>
        <Text>Question</Text>
        <TextInput
          ref={input => { textInputQuestion = input }}
          returnKeyType={"next"}
          blurOnSubmit={false}
          multiline={true}
          style={styles.textInput}
          placeholder="Enter with a question."
          onSubmitEditing={() => {
            textInputAnswer.focus()
          }}
          onChangeText={text => textInputQuestion.value = text}
          defaultValue={defaultQuestion}
        />
      </View>
      <View style={styles.containerTextInput}>
        <Text>Answer</Text>
        <TextInput
          ref={input => { textInputAnswer = input }}
          returnKeyType={"done"}
          blurOnSubmit={true}
          multiline={true}
          style={styles.textInput}
          placeholder="Enter the answer to the question."
          onChangeText={text => textInputAnswer.value = text}
          defaultValue={defaultAnswer}
        />
      </View>
      <View style={styles.buttonsContainer}>
        <TouchableOpacity
          style={[buttonStyle, styles.buttonSaveAsCorrect]}
          onPress={() => parseFields(true)}>
          <Text style={styles.textButtons}>Save as Correct</Text>
        </TouchableOpacity>
        <TouchableOpacity
          style={[buttonStyle, styles.buttonSaveAsIncorrect]}
          onPress={() => parseFields(false)}>
          <Text style={styles.textButtons}>Save as Incorrect</Text>
        </TouchableOpacity>
      </View>
    </KeyboardAvoidingView>
  )
}

const styles = StyleSheet.create({
  container: {
    flexDirection: 'column',
    flex: 1,
    marginBottom: 10
  },
  containerTextInput: { height: '30%' },
  textInput: { fontSize: 24, height: '90%', width: '100%' },
  buttonsContainer: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    flexWrap: 'wrap',
    marginBottom: 5
  },
  buttonSaveAsCorrect: {
    backgroundColor: green,
    margin: 5
  },
  buttonSaveAsIncorrect: {
    backgroundColor: red,
    margin: 5
  },
  textButtons: {
    color: white,
    fontSize: 24
  },

});