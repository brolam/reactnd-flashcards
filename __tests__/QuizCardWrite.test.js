import React from 'react'
import QuizCardWrite from '../components/QuizCardWrite'

test('renders without crashing', () => {
  const wrapper = shallow(<QuizCardWrite />);
  expect(wrapper).toMatchSnapshot();
});

test('onSave event expect an answer correct', () => {
  const onSave = jest.fn()
  const quizCardWrite = mount(
    <QuizCardWrite
      onSave={onSave}
    />);
  const textInputQuestion = quizCardWrite.find('TextInput').at(0)
  const textInputAnswer = quizCardWrite.find('TextInput').at(2)
  const buttonSaveCorrect = quizCardWrite.find('TouchableOpacity').at(0)
  textInputQuestion.props().onChangeText('One Question')
  textInputAnswer.props().onChangeText('One Answer')
  buttonSaveCorrect.props().onPress()
  expect(onSave).toHaveBeenCalledWith('One Question', 'One Answer', true);
});
