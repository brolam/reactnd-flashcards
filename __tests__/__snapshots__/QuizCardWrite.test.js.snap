// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<KeyboardAvoidingView
  behavior="padding"
  keyboardVerticalOffset={0}
  style={
    Array [
      Array [
        Object {
          "backgroundColor": "#fff",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginLeft": 10,
          "marginRight": 10,
          "marginTop": 8,
          "padding": 20,
          "shadowColor": "rgba(0, 0, 0, 0.24)",
          "shadowOpacity": 0.8,
        },
        Object {
          "borderRadius": 8,
          "shadowOffset": Object {
            "height": 3,
            "width": 0,
          },
          "shadowRadius": 3,
        },
      ],
      Object {
        "flex": 1,
        "flexDirection": "column",
        "marginBottom": 10,
      },
    ]
  }
>
  <View
    style={
      Object {
        "height": "30%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
    >
      Question
    </Text>
    <TextInput
      blurOnSubmit={false}
      defaultValue=""
      multiline={true}
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Enter with a question."
      returnKeyType="next"
      style={
        Object {
          "fontSize": 24,
          "height": "90%",
          "width": "100%",
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "height": "30%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
    >
      Answer
    </Text>
    <TextInput
      blurOnSubmit={true}
      defaultValue=""
      multiline={true}
      onChangeText={[Function]}
      placeholder="Enter the answer to the question."
      returnKeyType="done"
      style={
        Object {
          "fontSize": 24,
          "height": "90%",
          "width": "100%",
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "flexWrap": "wrap",
        "justifyContent": "space-around",
        "marginBottom": 5,
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
      style={
        Array [
          Array [
            Object {
              "alignItems": "center",
            },
            Object {
              "backgroundColor": "#fff",
              "borderColor": "#292477",
              "borderRadius": 3,
              "borderWidth": 1,
              "padding": 5,
              "paddingLeft": 25,
              "paddingRight": 25,
            },
          ],
          Object {
            "backgroundColor": "#3cb371",
            "margin": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 24,
          }
        }
      >
        Save as Correct
      </Text>
    </TouchableOpacity>
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
      style={
        Array [
          Array [
            Object {
              "alignItems": "center",
            },
            Object {
              "backgroundColor": "#fff",
              "borderColor": "#292477",
              "borderRadius": 3,
              "borderWidth": 1,
              "padding": 5,
              "paddingLeft": 25,
              "paddingRight": 25,
            },
          ],
          Object {
            "backgroundColor": "#ff0000",
            "margin": 5,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 24,
          }
        }
      >
        Save as Incorrect
      </Text>
    </TouchableOpacity>
  </View>
</KeyboardAvoidingView>
`;
